lib/Dialect/FIRRTL/CHIRRTLDialect.cpp

```diff
@@ -181,6 +181,62 @@ static void printMemoryPortOp(OpAsmPrinter &p, Operation *op,
   p.printOptionalAttrDict(op->getAttrs(), elides);
 }
 
+//===----------------------------------------------------------------------===//
+// MemoryDebugPortOp
+//===----------------------------------------------------------------------===//
+
+void MemoryDebugPortOp::build(OpBuilder &builder, OperationState &result,
+                              Type dataType, Value memory, StringRef name,
+                              ArrayRef<Attribute> annotations) {
+  build(builder, result, dataType, memory, name,
+        builder.getArrayAttr(annotations));
+}
+
+LogicalResult MemoryDebugPortOp::inferReturnTypes(
+    MLIRContext *context, Optional<Location> loc, ValueRange operands,
+    DictionaryAttr attrs, mlir::RegionRange regions,
+    SmallVectorImpl<Type> &results) {
+  auto inType = operands[0].getType();
+  auto memType = inType.dyn_cast<CMemoryType>();
+  if (!memType) {
+    if (loc)
+      mlir::emitError(*loc, "memory port requires memory operand");
+    return failure();
+  }
+  results.push_back(RefType::get(
+      FVectorType::get(memType.getElementType(), memType.getNumElements())));
+  return success();
+}
+
+void MemoryDebugPortOp::getAsmResultNames(
+    function_ref<void(Value, StringRef)> setNameFn) {
+  StringRef base = getName();
+  if (base.empty())
+    base = "memport";
+  setNameFn(getData(), (base + "_data").str());
+}
+
+static ParseResult parseMemoryDebugPortOp(OpAsmParser &parser,
+                                          NamedAttrList &resultAttrs) {
+  // Add an empty annotation array if none were parsed.
+  auto result = parser.parseOptionalAttrDict(resultAttrs);
+  if (!resultAttrs.get("annotations"))
+    resultAttrs.append("annotations", parser.getBuilder().getArrayAttr({}));
+  return result;
+}
+
+/// Always elide "direction" and elide "annotations" if it exists or
+/// if it is empty.
+static void printMemoryDebugPortOp(OpAsmPrinter &p, Operation *op,
+                                   DictionaryAttr attr) {
+  // "direction" is always elided.
+  SmallVector<StringRef> elides;
```

Comment:
This doesn't elide "direction", please sync comment and code.

And consider `SmallVector<StringRef,1>`, no reason to allocate more if it is max 1 element, but not a big deal.

Reply:
Thanks for spotting this. Fixed.